Connection pool refactoring

NiteshKant opened the pull request:
Made the following changes to ChannelPool:
- Improved the abstraction earlier provided by ChannelPool by converting it to a "ConnectionPool" to pool ObservableConnection. This fits in better with the rest of the code.
- Modified the interface to "assume" that a pool connects to one server. This also fits in better with the existing client abstraction.
- Provided a shutdown method on the pool. (Issue #80, #81)
- Modified the pool stats by not storing any monotonically increasing values as they overflow. Instead provided various callbacks which can be used to infer these stats.
- Removed the use of semaphores as they are easily replaced by an AtomicInteger. Also modified the method to update the max connections limit as atomic increment/decrements rather than a set, which is hard to make thread-safe and consistent without synchronization. (Discussed partially under issue #83)
- Fixed issue #82, #76 and #79

There are three pending tasks after this change:
- Provide a "cleanup thread" (using a threadpool) to cleanup idle connections.
- Provide a composite pool stats object encompassing multiple connection pools.
- Provide another PoolLimitDeterminationStrategy for max connections across pools which can be used to enforce connection limits over a pool of servers.

Commit:
Connection pool refactoring.

This is still WIP. The following are the todo:

1) Idle connection cleanup thread is not implemented.
2) More tests need to be written and the older one needs to be fixed.

Commit:
Connection Pool refactoring.

* Modified Connection pool interface to also send PipelineConfigurator to acquire(). This is required as the pipelineConfigurator is modified by the client to accommodate all RxNetty specific handlers. Storing the pipeline configurator provided by the user does not work for the pool.
* The shutdown of eventloop during client shutdown will shutdown the shared eventloop (by default). Added ref counting on the shared eventloop group, so it only shutsdown if there are no more references.
* Removed shutdown() call from waitTillShutdown() in RxServer as it is used as a waiting mechanism for an external shutdown for a server.
* Enabled ClientRequestResponseConverter to be reused by resetting the contentSubject.
* Added more tests and fixed the existing tests.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #12](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/12/) FAILURE
Looks like there's a problem with this pull request

Commit:
Fixing some tests which were too sensitive to timeout values.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #13](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/13/) FAILURE
Looks like there's a problem with this pull request

Commit:
Implemented a background cleanup task for idle connections.

Modified the mock RequestProcessor to not sleep in the request processing as that blocks the eventloop and hence produces unnecessary delays in the parallel executing tests.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #14](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/14/) FAILURE
Looks like there's a problem with this pull request

Commit:
Added a composite pool stats impl.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #15](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/15/) SUCCESS
This pull request looks good

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/client/PooledConnection.java:
```diff
@@ -0,0 +1,99 @@
+        Boolean discardConn = getChannelHandlerContext().channel()
+                              .attr(ClientRequestResponseConverter.DISCARD_CONNECTION).get();
+
+        if (!getChannelHandlerContext().channel().isActive() || null != discardConn && discardConn) {
```
can be simplified as discardConn == Boolean.TRUE

NiteshKant replied:
Sure, thanks!

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/channel/ObservableConnection.java:
```diff
@@ -61,34 +61,40 @@ public boolean isCloseIssued() {
+        ReadTimeoutPipelineConfigurator.removeTimeoutHandler(getChannelHandlerContext().pipeline());
+    }
+
+    protected Observable<Void> _closeChannel() {
```
why the "_" in the method name? Does it signal anything special?

NiteshKant replied:
Yeah I tend to add _ to non-public methods.

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java:
```diff
@@ -0,0 +1,262 @@
+        }
+        try {
+            stateChangeListener.onReleaseAttempted();
+            if (isShutdown.get()) {
```
some logic is repeated in the if-elseif-else. Can be simplified as

```
if (!isShutdown.get() && connection.isUsable()) {
    ...
} else {
   ...
}
```

NiteshKant replied:
Yes correct. will do.

allenxwang commented:
I would really like to see some JUnit test for usage of CompositeStrategy since it will be heavily used by Ribbon and any client that utilizes a pool that can connect to different servers.

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/client/CompositeStrategy.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * @author Nitesh Kant
+ */
+public class CompositeStrategy extends CompositePoolStateChangeListener implements PoolLimitDeterminationStrategy {
```
Possibly rename to CompositePoolLimitDeterminationStrategy to make it more obvious.

NiteshKant replied:
Sure

Commit:
Few more changes to the API.

1) Eliminated PoolStateChangeListener for an Observer. The pool now uses Observable to publish the state change events.
2) Added way to retrieve PoolStats and subscribe to change events from RxClient. This is the only way to access the pool. The reason why connection pool instance is not exposed from the RxClient is that it has write operations too.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #16](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/16/) FAILURE
Looks like there's a problem with this pull request

Commit:
Incorporated review comments by @awang

1) Minor code cleanups
2) Added test case for PoolLimitDeterminationStrategy. There were bugs in it :)

Commit:
Fix to PooledConnection.

The lastReturnToPoolTimeMillis was not getting updated as the Observable returned from close() was never subscribed to.
Since, release is synchronous, we do not need finallyDo() to update this stats.

NiteshKant commented:
@allenxwang I have added tests for PoolLimitDeterminationStrategy's various implementation. Thanks as there was actually a bug in there :)

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #17](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/17/) FAILURE
Looks like there's a problem with this pull request

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/client/MaxConnectionsBasedStrategy.java:
```diff
@@ -0,0 +1,73 @@
+         * This is just a more semantically correct implementation with similar performance characterstics as
+         * incrementAndGet()
+         */
+        for (;;) {
```
I am wondering in high concurrency, some requests will end up having to spend quite some time in the loop. Or, theoretically, a thread can never exist the loop - just before limitEnforcer.compareAndSet(currentValue, newValue) for that request is going to be executed, the thread is suspended and another thread gets a chance to modify limitEnforcer.

NiteshKant replied:
Yes but unfortunately, implementing atomic increments based on CAS are always like that. For reference if you look at jdk Semaphore implementation for tryAcquire() you will see similar code:

```
            for (;;) {
                int available = getState();
                int remaining = available - acquires;
                if (remaining < 0 ||
                    compareAndSetState(available, remaining))
                    return remaining;
            }
```

or AtomicInteger.incrementAndGet()

```
        for (;;) {
            int current = get();
            int next = current + 1;
            if (compareAndSet(current, next))
                return next;
        }
```

I am open to suggestions if you have a better way to do this.

allenxwang replied:
Can we fail fast if we have spend too many cycles in the loop, like 1000?

NiteshKant replied:
I think the correct terminology will be "fail slow" considering 1000 iterations :)

We surely can add that but that would mean that 1000 time concurrently at the same time two threads ended up in the same point. Specifically that this kind of loops are all over the Atomic package, adding this safety net seems unnecessary to me.

Commit:
Using a different eventloop group for ConnectionPoolTest as the close notifications get delayed in some build machines due to shared eventloop group.

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #18](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/18/) FAILURE
Looks like there's a problem with this pull request

Commit:
Attempt to fix the flaky test (only in cloudbees)

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #19](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/19/) SUCCESS
This pull request looks good

NiteshKant commented:
@allenxwang are you comfortable with me merging this now?

Commit:
Introducing PoolStatsProvider to aid providing an external PoolStats implementation to the pool.

allenxwang commented on rx-netty/src/test/java/io/reactivex/netty/client/PoolLimitStrategyTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    public void testCompositeStrategy() throws Exception {
```
I am wondering if we should write another test case that test the normal situation that the global permits are greater than the local permits. Also some tests showing that max connections are resettable would be good.

NiteshKant replied:
Sure, done.

allenxwang commented:
I feel it make sense to have a pool sub sub package under client since pool is not a required component for client.

allenxwang commented:
Do you have an example I can I subscribe to pool state change without the direct access to pool object itself? Is it through RxClientImpl.stateChangeObservable()? What if somebody creates the client using the builder and does not have access to the impl class?

allenxwang commented on rx-netty/build.gradle:
```diff
@@ -38,6 +38,8 @@ dependencies {
     compile 'io.netty:netty-codec-http:4.0.14.Final'
     compile 'com.netflix.rxjava:rxjava-core:[0.17,)'
+    compile 'org.slf4j:slf4j-api:1.7.6'
+    compile 'com.netflix.numerus:numerus:1.+'
```
should be testCompile?

NiteshKant replied:
Why? numerus is used for LongAdder in pool stats. slf4j is used for logging

NiteshKant commented:
> > > I feel it make sense to have a pool sub sub package under client since pool is not a required component for client.

How will having a different package name do anything with it being required or not?
The reason why it isn't a different package is that they share some package private classes/methods which are internal implementation detail.

NiteshKant commented:
> > > Do you have an example I can I subscribe to pool state change without the direct access to pool object itself? Is it through RxClientImpl.stateChangeObservable()? What if somebody creates the client using the builder and does not have access to the impl class?

RxClient interface provides this method (by virtue of extending it from PoolInsightProvider)

Commit:
Added more tests for the strategy

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #20](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/20/) FAILURE
Looks like there's a problem with this pull request

allenxwang commented on rx-netty/src/main/java/io/reactivex/netty/client/PoolInsightProvider.java:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @return An {@link Observable} emitting all state change events to the pool.
+     */
+    Observable<StateChangeEvent> stateChangeObservable();
```
Can we rename this poolStateChangeObservable and PoolStateChangeEvent? When this method is invoked on RxClient, it will be confusing what state change is about.

NiteshKant replied:
Since StateChangeEvent is an inner class (enum) inside PoolInsightProvider, I thought it was indicative enough with a shorter name.
Anyways, I have renamed it.

Commit:
Renaming PoolInsightProvider.StateChangeEvent to PoolInsightProvider.PoolStateChangeEvent

cloudbees-pull-request-builder commented:
[RxNetty-pull-requests #22](https://netflixoss.ci.cloudbees.com/job/RxNetty-pull-requests/22/) SUCCESS
This pull request looks good
